teachertool/src/types/criteriaParameters.ts

```diff
@@ -0,0 +1,139 @@
+import { Strings } from "../constants";
+import { logError } from "../services/loggingService";
+import { ErrorCode } from "./errorCode";
+
+export type CriteriaParameterType = "string" | "longString" | "number" | "block" | "system";
+
+export interface CriteriaParameterValidationResult {
+    valid: boolean;
+    message?: string;
+}
+
+// Represents a parameter definition in a catalog criteria.
+export class CriteriaParameter {
+    name: string;
+    type: CriteriaParameterType;
+    default: string | undefined;
+    paths: string[]; // The json path(s) to update with the parameter value in the catalog criteria.
+
+    constructor(name: string, type: CriteriaParameterType, paths: string[], defaultValue?: string) {
+        this.name = name;
+        this.type = type;
```

Comment:
The way inheritance is setup here feels very C#-like. For example, `type` is assigned at runtime. This doesn't allow the code to leverage TypeScript's type system. Assigning a literal value to `type` in the declaration itself would alleviate the need for all the typecasting in `createSpecificParameter`.

Reply:
Something like this should get you going in that direction:
```ts
export abstract class CriteriaParameter {
    name: string;
    abstract type: CriteriaParameterType;
```

then

```ts
export class StringParameter extends BaseStringParameter {
    type = "string";
```

Reply:
(edited to add `readonly`)

Reply:
Per our conversation, I've removed the classes and just gone with types and a big switch/case within a global function for validation. I also moved validation functions into a separate file, so the types file is just types again.